### PR TITLE
Make timestamp authoritative for master information

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description
This PR makes it explicit in all the places and `shard_sync` that the timestamp for the master term is the authoritative source of information for who is the master. We sync the information between the tablet record and shard record based on the timestamp.

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletmanager/shard_sync.go [hunk 1]
```diff
@@ -154,18 +156,24 @@ func (tm *TabletManager) shardSyncLoop(ctx context.Context, notifyChan <-chan st
 // If the shard record indicates a new master has taken over, this returns
 // success (we successfully synchronized), but the returned masterAlias will be
 // different from the input tablet.Alias.
-func syncShardMaster(ctx context.Context, ts *topo.Server, tablet *topodatapb.Tablet, masterTermStartTime time.Time) (masterAlias *topodatapb.TabletAlias, err error) {
+func syncShardMaster(ctx context.Context, ts *topo.Server, tablet *topodatapb.Tablet, masterTermStartTime time.Time) (masterAlias *topodatapb.TabletAlias, haveOutdatedInfo bool, err error) {
 	ctx, cancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
 	defer cancel()
 
 	var shardInfo *topo.ShardInfo
+	haveOutdatedInfo = false
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/shard_sync.go:164]: This confused me at first. Can you rename this to `shouldDemote`?
It is not that the tablet has "outdated" info. It is that the tablet's earlier timestamp means that it is no longer the primary.

**Review by deepthi — APPROVED** [review]
Looks good. We can merge once the review comments are addressed.

File: go/vt/vtctl/grpcvtctldserver/topo.go [hunk 2]
```diff
@@ -217,7 +218,7 @@ func deleteTablet(ctx context.Context, ts *topo.Server, alias *topodatapb.Tablet
 			}
 
 			si.MasterAlias = nil
-
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtctl/grpcvtctldserver/topo.go:220]: done

File: go/vt/vttablet/tabletmanager/shard_sync.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletmanager/shard_sync.go]: changed

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**